corehq/apps/app_manager/suite_xml/sections/entries.py

```diff
@@ -844,150 +869,64 @@ def get_manual_datum(action_, parent_filter_=''):
-        if parent_datums:
-            # we need to try and match the datums to the root module so that
-            # the navigation on the phone works correctly
-            # 1. Add in any datums that don't require user selection e.g. new case IDs
-            # 2. Match the datum ID for datums that appear in the same position and
-            #    will be loading the same case type
-            # see advanced_app_features#child-modules in docs
-            datum_ids = {d.datum.id: d for d in datums}
-            index = 0
-            changed_ids_by_case_tag = defaultdict(list)
-            for this_datum_meta, parent_datum_meta in list(zip_longest(datums, parent_datums)):
-                if this_datum_meta:
-                    update_refs(this_datum_meta, changed_ids_by_case_tag)
-                if not parent_datum_meta:
-                    continue
-                if not this_datum_meta or this_datum_meta.datum.id != parent_datum_meta.datum.id:
-                    if not parent_datum_meta.requires_selection:
-                        # Add parent datums of opened subcases and automatically-selected cases
-                        datums.insert(index, attr.evolve(parent_datum_meta, from_parent=True))
-                    elif this_datum_meta and this_datum_meta.case_type == parent_datum_meta.case_type:
-                        append_update(changed_ids_by_case_tag,
-                                      rename_other_id(this_datum_meta, parent_datum_meta, datum_ids))
-                        append_update(changed_ids_by_case_tag,
-                                      get_changed_id(this_datum_meta, parent_datum_meta))
-                        this_datum_meta.datum.id = parent_datum_meta.datum.id
-                index += 1
+    def add_parent_datums(self, datums, module):
+        parent_datums = self._get_first_forms_datums(module.root_module)
+        if not parent_datums:
+            return datums
+
+        # we need to try and match the datums to the root module so that
+        # the navigation on the phone works correctly
+        # 1. Add in any datums that don't require user selection e.g. new case IDs
+        # 2. Match the datum ID for datums that appear in the same position and
+        #    will be loading the same case type
+        # see advanced_app_features#child-modules in docs
+        datum_ids = {d.id: d for d in datums}
+        datums_by_case_tag = _get_datums_by_case_tag(datums)
+
+        def set_id(datum_meta, new_id):
+            case_tag = getattr(datum_meta.action, 'case_tag', 'basic')
+            _update_refs(datums_by_case_tag[case_tag], datum_meta.datum.id, new_id)
+            datum_meta.datum.id = new_id
+
+        ret = []
+        datums_remaining = list(datums)
+        for parent_datum_meta in parent_datums:
+            if parent_datum_meta.id in datum_ids:
+                datum = datum_ids[parent_datum_meta.id]
+                set_id(datum, f'{datum.id}_{datum.case_type}')
```

Comment:
I expected this change to break the `BasicModuleAsChildTest.test_child_module_parent_no_parent_select_same_case_type` test. Debugging locally I see that this first renames `case_id` to `case_id_gold_fish` and then on line 926 it renames it back again which is why the test doesn't break. 

I'm not sure how I feel about that but perhaps just a comment would be sufficient:
```suggestion
            if parent_datum_meta.id in datum_ids:
                # We assume that a conflict is unwanted and rename. This may get undone below.
                datum = datum_ids[parent_datum_meta.id]
                set_id(datum, f'{datum.id}_{datum.case_type}')
```

Reply:
I agree - I'm not crazy about it as-is.  I just tried a few variations, but wasn't able to get anything that looked better and still worked.  It'd be nice to have this only iterate over datum IDs remaining - it's weird that it could in theory mutate datums that have already been added to `ret`.  I think my ideal would be to have this pulled out of the loop and have a separate block that checks for ID conflicts when adding the remaining datums.  Anyways, I'm punting for now and just adding that comment.